### DISCUSSION TITLE
How use 'A is the immediate dependent of B.' operator in Spacy DependencyMatcher

### DISCUSSION THREAD

**Discussion by OscarArmas** [explosion/spaCy#8803] [Help: Coding & Implementations] [answered]
I want a pattern that matches the word 'company' and then selects the immediate dependent with the DEP 'dobj', my text is:

![image](https://user-images.githubusercontent.com/37764897/126885574-34098196-9978-48a1-b433-02138410d240.png)

**The pattern and code is:**

```
import spacy
from spacy.matcher import DependencyMatcher
nlp = spacy.load("en_core_web_sm")
matcher = DependencyMatcher(nlp.vocab)

pattern = [
    {
        "RIGHT_ID": "anchor_founded",
        "RIGHT_ATTRS": {"ORTH": "company"}
    },
    {
        "LEFT_ID": "anchor_founded",
        "REL_OP": "<",
        "RIGHT_ID": "founded_subject",
        "RIGHT_ATTRS": {"DEP": "dobj"},
    },
]

matcher.add("FOUNDED", [pattern])
doc = nlp("Carl has founded a cosmetic company.")
matches = matcher(doc)
```

I think with the above code the output has to be  ['company', 'founded'], but doesn't match anything, do I using correctly the operator '<'?.

**Comment by polm** [comment, marked as answer]
You're understanding the operator correctly but the way you're using it is a bit weird. 

If you have a sentence like "John founded Acme", "John" is the subject and "Acme" is the object. Because you are using the "dobj" dep, you have no subject in your rules, even though your right id is "founded_subject". 

What your second rule matches is:

- the parent of "company"
- that also has DEP of "dobj" 

Assuming that "company" is an object, this doesn't make any sense - "company" will have the "dobj" relation. 

Here's a modified version that gives your expected output. Note that while not strictly necessary, it usually makes things easier if you make the verb the anchor, since it tends to be higher up in the tree.

```
import spacy
from spacy.matcher import DependencyMatcher
nlp = spacy.load("en_core_web_sm")
matcher = DependencyMatcher(nlp.vocab)

pattern = [
    {
        "RIGHT_ID": "anchor_founded",
        "RIGHT_ATTRS": {"ORTH": "founded"}
    },
    {
        "LEFT_ID": "anchor_founded",
        "REL_OP": ">",
        "RIGHT_ID": "founded_object",
        "RIGHT_ATTRS": {"DEP":"dobj"},
    },
]

matcher.add("FOUNDED", [pattern])
doc = nlp("Carl has founded a cosmetic company.")
matches = matcher(doc)

for match in matcher(doc):
    print(match)
```
